transport/grpc/stream.go

```diff
@@ -0,0 +1,214 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/atomic"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/yarpcerrors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+type serverStream struct {
+	ctx     context.Context
+	reqMeta *transport.RequestMeta
+	stream  grpc.ServerStream
+}
+
+func newServerStream(ctx context.Context, reqMeta *transport.RequestMeta, stream grpc.ServerStream) *serverStream {
+	return &serverStream{
+		ctx:     ctx,
+		reqMeta: reqMeta,
+		stream:  stream,
+	}
+}
+
+func (ss *serverStream) Context() context.Context {
+	return ss.ctx
+}
+
+func (ss *serverStream) RequestMeta() *transport.RequestMeta {
+	return ss.reqMeta
+}
+
+func (ss *serverStream) SendMsg(m *transport.StreamMessage) error {
+	// TODO pool buffers for performance.
+	msg, err := ioutil.ReadAll(m)
+	_ = m.Close()
+	if err != nil {
+		return err
+	}
+	return toYARPCStreamError(ss.stream.SendMsg(msg))
+}
+
+func (ss *serverStream) RecvMsg() (*transport.StreamMessage, error) {
+	// TODO used pooled buffers for performance.
+	var msg []byte
+	if err := ss.stream.RecvMsg(&msg); err != nil {
+		s, ok := status.FromError(err)
+		if ok && s.Code() == codes.Canceled && s.Message() == context.Canceled.Error() {
+			// GRPC Race condition on end of a stream, sometimes it returns a context cancelled error
+			return nil, io.EOF
+		}
+		return nil, toYARPCStreamError(err)
+	}
+	return &transport.StreamMessage{ReadCloser: ioutil.NopCloser(bytes.NewReader(msg))}, nil
+}
+
+func (ss *serverStream) SetResponseMeta(respMeta *transport.ResponseMeta) {
+	if respMeta == nil {
+		return
+	}
+	md := metadata.New(nil)
+
+	// This can fail for validation reasons, we should probably set an error on
+	// the metadata if that's the case?
+	_ = addApplicationHeaders(md, respMeta.Headers)
```

Comment:
I'm returning errors in https://github.com/yarpc/yarpc-go/pull/1349, this might be a good thing to copy.

Reply:
Especially now that #1349 is merged :-)